chanfitness/chaneventstore.go

```diff
@@ -0,0 +1,308 @@
+	// stream of channel events.
+	ChannelEvents func() (*subscribe.Client, error)
+
+	// PeerEvents provides a subscription client which provides a stream of
+	// peer online/offline events.
+	PeerEvents func() (*subscribe.Client, error)
+
+	// Get open channels provides a list of existing open channels
+	// which is used to populate the ChannelEventStore with a set of channels
+	// on startup.
+	GetOpenChannels func() ([]*channeldb.OpenChannel, error)
+}
+
+// NewChannelEventStore initializes an event store with the config provided. Note
+// that this function does not start the monitoring process, Start() must be
+// called.
+func NewChannelEventStore(config *Config) *ChannelEventStore {
+	store := &ChannelEventStore{
+		cfg:      config,
+		channels: make(map[uint64]*chanEventLog),
+		peers:    make(map[[33]byte]bool),
+		quit:     make(chan struct{}),
+	}
+
+	return store
+}
+
+// Start adds all existing open channels to the event store and starts monitoring
+// for channel opened/closed events so that it can add and remove channels as
+// appropriate.
+func (c *ChannelEventStore) Start() error {
+	// Create a subscription to channel events.
+	channelClient, err := c.cfg.ChannelEvents()
+	if err != nil {
+		return err
+	}
+
+	// Create a subscription to peer events.
+	peerClient, err := c.cfg.PeerEvents()
+	if err != nil {
+		return err
+	}
+
+	// All events that need to be monitored are piped into this channel.
+	// This is done to assist mocking for testing.
+	events := make(chan interface{})
+
+	// Consume all events piped into the events channel.
+	c.wg.Add(1)
+	go c.monitorChannelEvents(events)
+
+	// This goroutine pipes all events into a single events channel, closing the
+	// channel and exiting if one of the event subscriptions fails, or the store
+	// is given the instruction to exit.
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+
+		// Cancel subscriptions on exit.
+		defer channelClient.Cancel()
+		defer peerClient.Cancel()
+
+		for {
+			select {
+			case e := <-channelClient.Updates():
+				events <- e
+
+			case e := <-peerClient.Updates():
+				events <- e
+
+			// Close the events channel and exit if one of the subscriptions
+			// fails, or the store receives the instruction to quit.
+			case <-c.quit:
+			case <-channelClient.Quit():
+			case <-peerClient.Quit():
+				close(events)
+				return
+			}
+		}
+	}()
+
+	// Add the existing set of channels to the event store. This is required
+	// because channel events will not be triggered for channels that exist
+	// at startup time.
+	channels, err := c.cfg.GetOpenChannels()
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Adding %v channels to event store", len(channels))
+
+	for _, ch := range channels {
+		var peerKey [33]byte
+		copy(peerKey[:], ch.IdentityPub.SerializeCompressed())
+
+		// Add existing channels to the channel store with an initial peer
+		// online or offline event.
+		c.addChannel(ch.ShortChanID().ToUint64(), peerKey)
```

Comment:
Should this be done before the goroutines are started?

Reply:
There's a chance that we miss a channel if this happens before we start monitoring
1. Add all channels from DB
2. Channel is opened and notification triggers
3. We start subscribing to events and have missed the channel open

Any reason for having it above that I'm missing? I also had it above to start because it looked better until Joost pointed this ^ out to me

Reply:
As long as the subscriptions is made before adding channels from the DB i think it should be fine. So order will be:
1. c.cfg.ChannelEvents() / PeerEvents
2. Add channels from db
3. start go routines

I can think of two good reasons to doing it in this order:
1. if `GetOpenChannels` returns an error, then we don't have to quit the goroutines before returning.
2. By adding the db channels before starting the go routines, we ensure we know about them before handling events.

Reply:
ah got it :) that's a better order

Reply:
I think the subscription to channel/peer events is processed asynchronously. The call to subscribe returning, doesn't seem to guarantuee that it has happened.

Reply:
Ok, my mistake. It is happening synchronously. Nvm